File: lib-walle-reader/src/main/java/cocos/creator/walle/ExtraInfoReader.java
package cocos.creator.walle;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public final class ExtraInfoReader {

    private ExtraInfoReader() {
        super();
    }

    /**
     * easy api for get extra info.<br/>
     *
     * @param apkFile apk file
     * @return null if not found
     */
    public static ExtraInfo get(final File apkFile) {
        final Map<String, String> result = getMap(apkFile);
        if (result == null) {
            return null;
        }
        return new ExtraInfo(result);
    }

    /**
     * get extra info by map
     *
     * @param apkFile apk file
     * @return null if not found
     */
    public static Map<String, String> getMap(final File apkFile) {
        try {
            final String rawString = getRaw(apkFile);
            if (rawString == null) {
                return null;
            }
            final JSONObject jsonObject = new JSONObject(rawString);
            final Iterator keys = jsonObject.keys();
            final Map<String, String> result = new HashMap<String, String>();
            while (keys.hasNext()) {
                final String key = keys.next().toString();
                result.put(key, jsonObject.getString(key));
            }
            return result;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * get raw string from block
     *
     * @param apkFile apk file
     * @return null if not found
     */
    public static String getRaw(final File apkFile) {
        return  PayloadReader.getString(apkFile, ApkUtil.EXTRA_INFO_BLOCK_ID);
    }
}

File: app-sdk/src/main/java/code/sdk/zfutil/ExtraInfoReader.kt
package code.sdk.zfutil

import org.json.JSONException
import org.json.JSONObject
import java.io.File

object ExtraInfoReader {
    /**
     * easy api for get extra info.<br></br>
     *
     * @param apkFile apk file
     * @return null if not found
     */
    operator fun get(apkFile: File?): ExtraInfo? {
        val result = getMap(apkFile) ?: return null
        return ExtraInfo(result)
    }

    /**
     * get extra info by map
     *
     * @param apkFile apk file
     * @return null if not found
     */
    fun getMap(apkFile: File?): Map<String, String>? {
        try {
            val rawString = getRaw(apkFile) ?: return null
            val jsonObject = JSONObject(rawString)
            val keys = jsonObject.keys()
            val result: MutableMap<String, String> = HashMap()
            while (keys.hasNext()) {
                val key = keys.next()
                result[key] = jsonObject.getString(key)
            }
            return result
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        return null
    }

    /**
     * get raw string from block
     *
     * @param apkFile apk file
     * @return null if not found
     */
    fun getRaw(apkFile: File?): String? {
        return PayloadReader.getString(apkFile, ApkPackageUtil.EXTRA_INFO_BLOCK_ID)
    }
}
